### PR TITLE
[mojo-stdlib] Add new `ComparableCollectionElement` trait and `index(list, value, start, end)` function APIs

### PR DESCRIPTION
**Pull request by arvindavoudi**
This pull request introduces the new `index` method to the `collections.List` struct within mojo-stdlib, mirroring the exact functionality found in Python's list, as detailed [here](https://docs.python.org/3/tutorial/datastructures.html).

This also allows for the specification of start and end parameters when searching for an item's index within a list, with both parameters accepting positive and negative numbers.

Also comprehensive suite of unit tests included meticulously designed to cover a wide range of scenarios for the `index` method's application.

### CODE REVIEW

File: stdlib/src/collections/list.mojo
Comment (JoeLoser) [inline stdlib/src/collections/list.mojo:356]: FYI issue 2188 seems to be resolved.  Can you uncomment this part and see if the `raise` works?

File: stdlib/src/collections/list.mojo
Comment (JoeLoser) [inline stdlib/src/collections/list.mojo:344]: **Suggestion** It may be good to include `normalized_start` as well as `len(self)`, i.e. the list size in the error message raised.  Ditto in the overload below.

File: stdlib/test/collections/test_list.mojo
Comment (JoeLoser) [inline stdlib/test/collections/test_list.mojo:296]: **Suggestion** You could specify `contains=<some_msg>` as a keyword argument to the `assert_raises` to verify the error message if you'd like.  This would be nice if you adopt my earlier comment about including the `normalized_start` and length of the list in the error message.

File: stdlib/test/collections/test_list.mojo
Comment (JoeLoser) [inline stdlib/test/collections/test_list.mojo:301]: **Question** Is this one correct?  I think in python,

```python
xs = [10, 20, 30, 40, 50]
xs.index(30, -4, 2)
```

will raise.

File: stdlib/test/collections/test_list.mojo
Comment (JoeLoser) [inline stdlib/test/collections/test_list.mojo:330]: **Question** Is this one correct as well?  Similar to my question about the test case above.

**Comment by JoeLoser** [comment]
> Additionally, it's worth noting that the `list.index` method in Python accepts the parameters `value`, `start`, and `end`. However, these are not treated as keyword arguments, meaning that while passing the `start` and `end` parameters is optional, if you wish to specify `end`, you must also pass `start`. This limitation informed my decision to overload the index method as you can observe.
> 
> Personally, I would prefer the flexibility to omit the `start` parameter, especially since the CPython API is gradually embracing keyword arguments for functions that previously did not support them. This is evidenced by recent developments, such as this [PR#106488](https://github.com/python/cpython/pull/106488) for Python 3.13, which aligns with this preference.
> 
> I'm eager to hear your thoughts on this matter. Thanks in advance.

Thanks for bringing this up.  This is interesting from an API design evolution standpoint.  Given that we already have keyword arguments, and modern Python is moving in this direction, I'd personally like to intersect with where they're going given where the Mojo language is at right now, rather than trying to be exactly source-compatible with older Python versions exclusively.  However, from a pragmatic "now" perspective, I'd recommend we keep the API as you have it, even though that's not using the keyword argument form.  I think we should raise this up into a broader API evolution guidelines talking point.  I'll think about the right forum to raise this in (whether internally or as a discussion externally) and go from there.  Let's not let the (lack of) a decision on that block this PR though.

FYI: @modularml/mojo-standard-library @jopamer

**Review by JoeLoser — COMMENTED** [review]
Generally looks good — please take a look at my comments though.  In addition, you'll need to rebase with the latest `nightly` branch and resolve the conflicts in one file.  Thanks for working on this!

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo:356]: Sure. I'm going to update it.

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo:344]: Great idea for improving error messages in Mojo, similar to the enhancements we've observed in CPython over recent years. I will implement this as well. Thanks

File: stdlib/test/collections/test_list.mojo
Comment (arvindavoudi) [reply stdlib/test/collections/test_list.mojo:301]: Um, It seems you misread the test case. We are not calling `index(30, -4, 2)`. Instead, the method is `index(30, -4)`, and then we use `assert_equal` to verify whether the result equals `2`. So the final code for that test is correct as below:

```mojo
assert_equal(test_list_a.index(30, start=-4), 2)
```

Also if you run your given code in Mojo, it will raise an error just like Python. I believe that the implemented `index` method does not have any issues here.

File: stdlib/test/collections/test_list.mojo
Comment (arvindavoudi) [reply stdlib/test/collections/test_list.mojo:330]: Yes, I think that's correct as I described above.

File: stdlib/test/collections/test_list.mojo
Comment (JoeLoser) [reply stdlib/test/collections/test_list.mojo:301]: Oh you're definitely right here. Sorry I misread the test case.

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo:344]: One quick question. In Python, if you give out-of-range start/end values, it does not raise any error concerning that bad given range. It just raises a value error in any case that it could not find the value. So, in Python, the code below works but in my implementation, it does not work:

```python
x = [10, 20, 30]
x.index(10, 0, 500) # Search for value; If does not exists, raise a Value Error. Don't care if start/end is out of range
```

So do you think we should keep this raising as it is? Or because of this incompatibility, it's better to drop it off?

I suddenly noticed about this so I personally prefer to have the same behavior in Mojo as well.

File: stdlib/test/collections/test_list.mojo
Comment (arvindavoudi) [reply stdlib/test/collections/test_list.mojo]: Done

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo]: I fixed it to use an appropriate approach. Neither losing clarity of the error messages you desired nor making incompatibility with Python's API.

**Comment by arvindavoudi** [comment]
Now I think we must wait for [this issue](https://github.com/modularml/mojo/issues/2238) to get resolved.

**Comment by JoeLoser** [comment]
> Now I think we must wait for [this issue](https://github.com/modularml/mojo/issues/2238) to get resolved.

Instead of blocking on a resolution to that issue, what do you think about making these free functions in the `list.mojo` file and the first argument is `List[T]` instead of the `self` argument now?  I'd recommend we move forward in this direction and can move them to be a member function once our traits are further along.

In general, anywhere where we need these trait refinements (like `__eq__` or `__str__` requirement), we could make them be free functions for now, like we do with `UnsafePointer` extension methods for example [here](https://github.com/modularml/mojo/blob/0a884b7941303e41d77e889101c34783c9e3a975/stdlib/src/memory/unsafe_pointer.mojo#L346).  FYI @gabrieldemarmiesse as I think you mentioned something on Discord recently about this sort of thing.

**Comment by gabrieldemarmiesse** [comment]
@arvindavoudi You can use this implementation as reference: https://github.com/modularml/mojo/pull/2323  I cause here a small parser bug, but this should not happen if you use this strategy to implement `List.index()`

Users will be able to call it with `__type_of(my_list).index(my_list, value, start, end)`

**Comment by arvindavoudi** [comment]
@JoeLoser, @gabrieldemarmiesse
Thanks for your concern. But one part is still ambiguous for me. In [issue 2263](https://github.com/modularml/mojo/issues/2263), you mentioned `count` implementation should be like below (which is very similar to `index`):

```mojo
def count[T: CollectionElement](arg: List[T], value: T):
   ...
```

But I didn't quite understand what approach we should take to be able to compare two values when they are both identified as CollectionElement, so then not comparable. In my PR this comparing part was specified as:

```mojo
if ((self.data + i).bitcast[T]()[]) == value:
```

I highly appreciate if you could clarify it for me as I can't really see any striking difference in here.

**Comment by gabrieldemarmiesse** [comment]
I think it was a small oversight on @JoeLoser 's part, you are right. `CollectionElement`'s types are not `EqualityComparable`, so you can't compare easily elements.

You can follow #2323  to see how to do it in practice, but I'll give you a summary. There are 3 main steps for you:

### Make a new trait that combines `CollectionElement` and `EqualityComparable`:
in mojo it would look like this:

```mojo
trait ComparableCollectionElement(EqualityComparable, CollectionElement):
    pass
```
### Create a static method on the List struct
Name the first argument `self` (so we won't have to rename it later on) and specify the type as corresponding to the `ComparableCollectionElement` trait. The signature will look like this:

```mojo
@staticmethod
fn index[U:ComparableCollectionElement](self: List[U], owned value: U, start: Int = 0) raises -> Int:
```
(I looked at the code in your PR as a reference).

### Use the static method
To use the method, since it's a staticmethod, you need to access the type of the list object you're working with. It will look like this:

```mojo
var my_list = List[Int](10, 11, 12)
var index_of_11 = __type_of(my_list).index(my_list, 11)
# note that in the future, we can remove @staticmethod and users will replace the call by "var index_of_11 = my_list.index(11)"
# it will be easy to replace by searching "__type_of" in the project
```

Is this enough to get you started? If you have any other questions feel free to ask :)

**Comment by JoeLoser** [comment]
As @gabrieldemarmiesse pointed out, that was just a small oversight on my part — sorry for the confusion (I've updated the GitHub issue to avoid future confusion).  We do need a composite trait to encompass that equality comparable requirement.  Mojo doesn't have in-place "anonymous composition" of traits, like (bikeshed syntax) 

```
def my_function[T: EqualityComparable + CollectionElement](value: T):
  pass
```

Instead, you have to create a named trait for each "anonymous" composition, e.g.

```
trait ComparableCollectionElement(EqualityComparable, CollectionElement):
    pass
```

This isn't ideal, but workable in the short-term until traits are further along in the compiler.  In this case, to do that "refinement" to ensure the given type can be guaranteed to have the `__eq__` function you want to use in your algorithm, we need said trait.

I don't have strong preferences on whether we make it a static method or free function outside of the struct until we have conditional conformances.  😄 

The key point is:
- We won't have conditional conformance anytime soon necessarily, so don't block this PR on that being available
- The reworking of `CollectionElement` trait to include `EqualityComparable` is complicated and won't be a short-term thing, so again, don't block this PR on that

So, we'll get creative in order for the library to move forward until we can get those nicer things in the future.  I have plans to push on (2) a bit sometime in the coming month or two.

**Comment by arvindavoudi** [comment]
Thanks for your detailed clarification @gabrieldemarmiesse.
I rebased and implemented those two as free functions as @JoeLoser suggested.
I also decided to put the new `ComparableCollectionElement` trait beside `CollectionElement` in `builtin/value.mojo` so it must be globally accessible from now on..?

File: stdlib/src/collections/list.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/collections/list.mojo:574]: If I may suggest, making this a `@staticmethod` instead of a free function. A few reasons I can think of:

* it allows a user to call this method without importing anything.
* It allows `__type_of(my_list).index(my_list, "hello")` which is easy to search and replace when #1876 is fixed
* When #1876 is fixed, users and us stdlib authors won't have to change the imports
* When #1876 is fixed, we won't have to move the code into the struct, avoiding a big diff due to indentation (for free functions, the indentation is 0 spaces, for methods it's 4 spaces)
* For built-in structs, users don't import anything to use them. So it's not obvious where they should import the free function from: `from ??? import index`. For us, standard library devs, it's obvious what the import is, but for Mojo users, they'll have to look at the docs to know the correct import.
* In this PR, the `index` method is imported in `stdlib/src/collections/__init__.mojo`, what do we do when multiple collections have a method with the same name, like `__eq__` ?
* For methods common to multiple types, it's going to get confusing. Let's say the user needs `__eq__` on both a List and a Dict in the same file, then users will have to do 
```mojo
from collections.list import __eq__
from collections.dict import __eq__
```
and it's going to get confusing.

For all those reasons, I suggest going for the syntax `__type_of(my_list).index(my_list, "hello")` which will lead to less issues in the future.

You can take a look at the way to make this syntax happen in this PR: https://github.com/modularml/mojo/pull/2323

File: stdlib/src/collections/list.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/collections/list.mojo:436]:
```suggestion
            raise "ValueError: Element is not in list."
```

To mimic the error message of python.

File: stdlib/src/collections/list.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/collections/list.mojo:419]: I don't think you need to implement the whole logic twice. You have two options:

```mojo
@staticmethod
fn index[C: ComparableCollectionElement](self: List[C], owned value: C, start: Int = 0) raises -> Int:
    return Self.index(self, value^, start, len(self))
```
this will call the second overload where the full implementation is. Or you have a second option, implementing only one signature:

```mojo
@staticmethod
fn index[C: ComparableCollectionElement](self: List[C], owned value: C, start: Int = 0, end: Optional[Int] = None) raises -> Int:
    if end is None:
        end = len(self)
    ...
```

File: stdlib/src/collections/list.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/collections/list.mojo:434]: You can just do `return i` instead of the `break` here. That should still work, and you won't need `ret_val`

File: stdlib/src/collections/list.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/collections/list.mojo:432]:
```suggestion
            if self[i] == value:
```
woudln't that work too?

File: stdlib/test/collections/test_list.mojo
Comment (gabrieldemarmiesse) [inline stdlib/test/collections/test_list.mojo:355]: I think the comments are a bit too much here. Comments should be there to explain something being done implicitely in the code, or to provide context that the reader can't guess, or even explain complexe algorithms. Here the comments are explaining in english what the code does. I don't think they're useful. The comments separating each section are relevant though.

**Review by gabrieldemarmiesse — COMMENTED** [review]
Thank you for the PR, could you run "mojo format ./" before pushing the new changes? Thanks! Also note that you can contest all my comments, I'm not modular staff nor do I represent them in any way 😁

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo]: Done

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo]: You are right. This method was used before the resolution of issue 2188. Done

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo]: Done

File: stdlib/test/collections/test_list.mojo
Comment (arvindavoudi) [reply stdlib/test/collections/test_list.mojo]: Thanks. Done. It's now definitely cleaner.

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo]: Thanks for bringing this up. I was also looking forward to implementing this, but I first wanted to have an acceptable version of `index`. Now that it's approved, I also going to merge those methods two by the help of `optional` in Mojo.
This also means we can pass `end` parameter without the need for specifying `start`. So I also added some new tests to cover this scenario.

**Comment by arvindavoudi** [comment]
> could you run "mojo format ./" before pushing the new changes?

Sure. I run it but it seems there are more formatting issues available in `list.mojo` and `test_list.mojo` that are not related to this PR. I doubt whether to commit those in here or not.

File: stdlib/src/collections/list.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/collections/list.mojo:427]: That's a lot of logic in the same line, and can make it quite hard to read. Can you split it into multiple lines?

**Comment by gabrieldemarmiesse** [comment]
It seems that you have git conflicts here. You'll need to either merge or rebase the "nightly" branch into your branch and fix the conflicts. Afterwards, make sure you have the latest version of mojo nightly with `modular clean && modular install nightly/mojo`, then you'll be able to run `mojo format ./`. 

If you have a hard time resolving the conflicts, or that you're afrait of loosing your changes, you can add me with commit rights to https://github.com/arvindavoudi/mojo and I'll fix the conflicts for you :)

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo]: Ah, my bad. Done.

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo]: Done.

File: stdlib/test/collections/test_list.mojo
Comment (arvindavoudi) [reply stdlib/test/collections/test_list.mojo]: In regular Mojo scripts, when a function returns a value that is not assigned to a variable, a warning is displayed. To address this, I assigned all those to "_" but I removed them in the newly committed changes.

File: stdlib/src/collections/list.mojo
Comment (arvindavoudi) [reply stdlib/src/collections/list.mojo:478]: Um yeah. This suggests that if want to first check `value` availability in the list using `in`, this makes the algorithm iterate over the list two times if the value is available. One time for checking value availability and the second time for finding its index. In the first version which was not clean and simple at all, I thought these comments may lead to preventing future "refactors" that can corrupt the performance. Since it's much more understandable now, I also assume that's unnecessary, so I will remove them.

**Comment by arvindavoudi** [comment]
Dear @gabrieldemarmiesse. I fixed everything you mentioned + the formatting. I hope there is no crucial problem anymore, ready to get merged :)

**Comment by gabrieldemarmiesse** [comment]
Mmhhh the CI is still not passing. That's weird. It seems the formatting is still not okay, if you want I'll take a look and see if I can push the commits directly to your branch.

**Comment by arvindavoudi** [comment]
Sure thanks. I have added you as a collaborator.

File: stdlib/test/collections/test_list.mojo
Comment (arvindavoudi) [reply stdlib/test/collections/test_list.mojo]: We can see warnings in the GitHub actions job result.

**Comment by arvindavoudi** [comment]
@gabrieldemarmiesse. No need for any further changes, I fixed the remaining issues. Thanks a lot for your valuable cooperation during the submission of this PR.

**Comment by gabrieldemarmiesse** [comment]
Great! Thank you for your hard work there!

**Review by JoeLoser — APPROVED** [review]
✅🟣 This contribution has been merged 🟣✅

Hey @arvindavoudi,

Thanks so much for the contribution — love the tests!  🎉 🔥 

We're moving to a new infrastructure for merging contributions to Mojo (we're using a tool called [Copybara](https://github.com/google/copybara)), and your contribution has now been merged into our internal copy of the Mojo Standard Library. I've added the "merged-internally" label on this PR.

The changes in this PR will appear here in the mojo repo nightly branch when we do our next outbound synchronization at the time that the next Mojo nightly is released. That should happen later today.

Please let me know if you have any questions or concerns.

**Comment by ConnorGray** [comment]
Closing as this has been merged internally, per Joe's comment above.

File: stdlib/src/collections/list.mojo
Comment (ConnorGray) [inline stdlib/src/collections/list.mojo:424]: This PR is amazing, thank you for the contribution @arvindavoudi and all the help @gabrieldemarmiesse.

One minor question we could possible address in a follow-up PR: Does this `value` argument used for comparision need to be `owned`, or could use the default `borrowed` argument convention? I.e.:

```mojo
self: List[C], value: C, ....
```

Logically I think `borrowed` would make sense, since the element isn't mutated or consumed by this method.

Is there some other reason this methods needs its own unique copy of the `value` argument?

Comment (JoeLoser) [reply stdlib/src/collections/list.mojo:424]: I had the same thought this morning — something we should follow-up on.

Comment (ConnorGray) [reply stdlib/src/collections/list.mojo:424]: No worries :) And yep, we can take of this internally as well.

Once we sync our internal changes out again, it will be easier for a community member to make this change as well if they wanted to. Its a small fix though, so I might just do it real quick internally if no one else gets to it first :)

Comment (ConnorGray) [reply stdlib/src/collections/list.mojo:424]: This has been fixed by our new intern @Dan13llljws in [this commit here](https://github.com/modularml/mojo/commit/f2f9e77ecf9780af4ecdf28ca170ca9276f586a9) — welcome to the team @Dan13llljws, and congratulations on your first Mojo🔥 contribution! 🎉  😄

**Pull request closed without merging**
